### DISCUSSION TITLE
env variables are undefined on browser in production

### DISCUSSION THREAD

**Discussion by prateekvijayvergiya** [vercel/next.js#21938] [Help]
Hi, I am working with nextjs version "10.0.6" and have some env variables in my .env.local file. I am accessing them by using process.env.NEXT_PUBLIC_SERVICE_ID. It is working fine in my local machine but giving undefined on production.

This is how my .env.local looks like
```
NEXT_PUBLIC_SERVICE_ID=hello
NEXT_PUBLIC_TEMPLATE_ID=hello
NEXT_PUBLIC_USER_ID=hello
```

> Note - I have deployed my application on gh-pages. I tried accessing those variables by declaring an env object in next.config.js but it's also giving me undefined,

Any suggestions ?

**Comment by michrome** [comment]
Hey @prateekvijayvergiya have you seen the [gh-pages Hello World example](https://github.com/vercel/next.js/tree/canary/examples/gh-pages)?

You need to build and export on your machine, then push the output because GitHub Pages doesn't run these steps like [Vercel](https://vercel.com/) does. In the example, they're using a [deploy script](https://github.com/vercel/next.js/blob/canary/examples/gh-pages/package.json#L9) to do it.

To configure the environment, they're using [`env-config.js`](https://github.com/vercel/next.js/blob/canary/examples/gh-pages/env-config.js) and the  [babel-plugin-transform-define](https://www.npmjs.com/package/babel-plugin-transform-define) plugin via [.babelrc.js](https://github.com/vercel/next.js/blob/canary/examples/gh-pages/.babelrc.js).

**Reply by prateekvijayvergiya** [reply]
In `env.config.js` they are exposing the values of env variables which I don't want to.

**Reply by michrome** [reply]
By design, [environment variables prefixed with `NEXT_PUBLIC_` are exposed to the browser](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser). 

If you’ve a ‘secret’ env then you should only use it in either `getStaticProps` or `getStaticPaths`. Its value won’t be included in the JS bundle for the browser.

**Reply by prateekvijayvergiya** [reply]
No problem, thanks @michrome I resolved this thing. Issue was with my deployment file.
